### ISSUE TITLE
[tlul] Data on write responses

### ISSUE THREAD

**Issue by GregAC** [lowRISC/opentitan#14611]
For a response to a write the data in the response packet (the `d_data` field) is irrelevant and could be any arbitrary value. It follows that the data integrity (`data_intg` within `d_user`) is also irrelevant and in particular doesn't need to be the correct integrity for whatever arbitrary value `d_data` has.

A recently found Ibex bug where Ibex was checking data integrity in incoming write responses (https://github.com/lowRISC/ibex/pull/1797) has highlighted that within OT Ibex is receiving valid data integrity for whatever value is supplied on `d_data` for a write response (otherwise we'd have seen many instances of this bug with chip-level tests).

Do we place a specific value on `d_data` for write responses? From a quick look my suspicion is instead the last read value just ends up staying on `d_data` along with its valid integrity. I think blanking it to some specific value along with invalid integrity would be a better choice.

I don't think there's major security risk from allowing the last read value to be seen on the write response. Perhaps an attacker could have control over some U mode code and is aware some M mode code has just read a sensitive value and does a write to attempt to capture the data from the bus but this would require glitching Ibex to turn the store into a load and may also require writing to the same device/memory that the sensitive value was read from (which presumably will be blocked by PMP). Still this should be a modest change thus still worth doing.

@eunchan @msfschaffner @tjaychen @moidx any thoughts?

**Comment by eunchan** [comment]
Make `d_data` blank (all zero or all one) is good IMO. Invalid integrity, hum.. I think it still should have integrity with the blank data. And would be good to check the integrity (it is solely up to the host honestly).

**Comment by msfschaffner** [comment]
This is related to https://github.com/lowRISC/opentitan/issues/11036 which blanks data going from host to device. As far as I remember, we tie the integrity to a fixed bad value whenever the data is blanked (see https://github.com/lowRISC/opentitan/pull/11045#discussion_r820026237), so we should probably do that here as well for consistency.

Relevant code section:
https://github.com/lowRISC/opentitan/blob/b05280af49cb8e08b1a2c3f4716c5fc8fe66d6cb/hw/ip/tlul/rtl/tlul_socket_1n.sv#L140-L148

CC: @tjaychen @weicaiyang

**Comment by GregAC** [comment]
We should have brought this up in the silicon meeting, apologies managed to forget about it!

> I think it still should have integrity with the blank data. And would be good to check the integrity (it is solely up to the host honestly).

I favour the bad integrity as a host consuming the read data on a write response simply shouldn't happen. If we give it good integrity we might miss this happening and end up with a more mysterious bug further down the line.

Being consistent as @msfschaffner states also seems good.

**Comment by msfschaffner** [comment]
Ok, let's put this on the agenda for the next meeting then!

**Comment by cdgori** [comment]
@tjaychen asked me to take a look at this as well.  I know we just spoke about it on the silicon call.

100% agree that it should be a fixed-value return rather than last-read value, that change has to happen.

As far as integrity-good or integrity-bad for the response, I am slightly in favor of integrity-good, for the case 2 mentioned by @weicaiyang above - we effectively get more distributed glitch detection by having integrity-good maintained.

My only concern would be if what @GregAC mentioned ("host consuming the read data on a write response") is possible, without some corresponding (detected) command integrity failure - I'm assuming this is of the "glitch attack converts a write into a read" type of situation.  I don't know enough about the internal workings of the TLUL adapter and/or Ibex here to say if this is possible to happen in an undetected way, maybe some others could clarify?

If we choose integrity-good does that mean that we are doing something different in the TLUL adapter based on #11036 / #11045 and should we try to make things consistent?  Or is there a good reason for the inconsistency?

**Comment by GregAC** [comment]
@cdgori The outcome from the silicon meeting was that we should give a fixed-value with integrity good. @weicaiyang believes this is already done within the TL fabric (it's not last value read as I postulated) but I think it'd be good to positively identify where within the RTL this happens as I didn't see it myself when I did a cursory check.

This was chosen as it's thought this is what is already happening so is the minimum change. The potential thing we would change is revert my fix in Ibex so it continues to check data integrity in write responses, do we feel this is important?

>  I don't know enough about the internal workings of the TLUL adapter and/or Ibex here to say if this is possible to happen in an undetected way, maybe some others could clarify?

In ibex there's a single flop you could glitch that would cause a write response to write whatever's in its data payload to the register file (with the destination register being chosen by the bottom bits of the address offset for the store, just an artifact of instruction encoding). However the lockstep check would catch this (one core would signal an RF write where the other wouldn't).

There's nothing within the TLUL fabric you could glitch to make Ibex think it was a read response rather than a write response. That information is not encoded in the response.

**Comment by cdgori** [comment]
> This was chosen as it's thought this is what is already happening so is the minimum change. The potential thing we would change is revert my fix in Ibex so it continues to check data integrity in write responses, do we feel this is important?

It does benefit us to check the integrity whenever possible, this is the "distributed glitch detection" I referred to above, so if reverting that fix in Ibex is a reasonable thing to do (i.e. doesn't introduce substantial extra verification overhead/cases), then I would be in favor of reverting it such that integrity is always checked.  If the overhead is high, let's discuss further/with wider audience the potential benefit.

> > I don't know enough about the internal workings of the TLUL adapter and/or Ibex here to say if this is possible to happen in an undetected way, maybe some others could clarify?
> 
> In ibex there's a single flop you could glitch that would cause a write response to write whatever's in its data payload to the register file (with the destination register being chosen by the bottom bits of the address offset for the store, just an artifact of instruction encoding). However the lockstep check would catch this (one core would signal an RF write where the other wouldn't).

That seems ~adequate to me.  (In theory, this could generate a hamming weight sidechannel leak of potentially sensitive information - imagine writing all-1's to a location holding a key.  It's an argument for not handling sensitive material in Ibex directly, of course.)  Since an adversary powerful enough to do this kind of glitching can probably do other equally focused glitching that more-directly reveals the sensitive information I don't think there is much more needed to be done (i.e. converting single flop to MUBI or other techniques).

> There's nothing within the TLUL fabric you could glitch to make Ibex think it was a read response rather than a write response. That information is not encoded in the response.

Great, this is what I hoped.

**Comment by msfschaffner** [comment]
@GregAC Afaik, the error value on responses is not set in the fabric directly, but inside the `tlul_adapter_sram` and `tlul_adapter_reg` modules. The data is set to all-ones in case of an error or write response, see:

https://github.com/lowRISC/opentitan/blob/75e9c17354d9c01aa257b2e2996c1b635c5a003e/hw/ip/tlul/rtl/tlul_adapter_reg.sv#L94-L97

https://github.com/lowRISC/opentitan/blob/75e9c17354d9c01aa257b2e2996c1b635c5a003e/hw/ip/tlul/rtl/tlul_adapter_sram.sv#L264-L292

**Comment by GregAC** [comment]
PR now open in Ibex to add integrity checking on write responses back in: https://github.com/lowRISC/ibex/pull/1835

It also updates the documentation to make it clear this is done.

**Comment by msfschaffner** [comment]
Seems like this can be closed since the newest Ibex patch above has been re-vendored into OpenTitan.
Please reopen if there is still something left to do here.
